src/middlewared/middlewared/plugins/system_general/update.py

```diff
@@ -244,8 +271,59 @@ async def do_update(self, data):
 
         await self.middleware.call('service.start', 'ssl')
 
+        if rollback_timeout is not None:
+            self._original_datastore = original_datastore
+            self._rollback_timer = asyncio.get_event_loop().call_later(
+                rollback_timeout,
+                lambda: asyncio.ensure_future(self.rollback()),
+            )
+
+        if ui_restart_delay is not None:
+            await self.middleware.call('system.general.ui_restart', ui_restart_delay)
+
         return await self.config()
 
     @private
     def set_crash_reporting(self):
         CrashReporting.enabled_in_settings = self.middleware.call_sync('system.general.config')['crash_reporting']
+
+    @accepts()
+    @returns(Int('remaining_seconds', null=True))
+    async def checkin_waiting(self):
+        """
+        Determines whether or not we are waiting user to check-in the applied UI settings changes before they are rolled
+        back. Returns a number of seconds before the automatic rollback or null if there are no changes pending.
+        """
+        if self._rollback_timer:
+            remaining = self._rollback_timer.when() - asyncio.get_event_loop().time()
+            if remaining > 0:
+                return int(remaining)
+
+    @accepts()
+    @returns()
+    async def checkin(self):
+        """
+        After UI settings are saved with `rollback_timeout` this method needs to be called within that timeout limit
+        to prevent reverting the changes.
+
+        This is to ensure user verifies the changes went as planned and its working.
+        """
+        if self._rollback_timer:
+            self._rollback_timer.cancel()
+
+        self._rollback_timer = None
+        self._original_datastore = {}
+
+    @private
+    async def rollback(self):
+        if self._original_datastore:
+            await self.middleware.call(
+                'datastore.update',
+                self._config.datastore,
+                self._original_datastore['id'],
+                {k: v for k, v in self._original_datastore.items() if k.startswith('stg_gui')},
```

Comment:
Just curious but you're caching the entirety of the database but only rolling back these columns? Seems like that could cause confusion.

Reply:
Yes, I only roll back related columns; however, I don't see how caching the entire row would hurt anyone.